### PR TITLE
libprom: init at 0.1.1

### PR DESCRIPTION
**Pull request by cfsmp3**
###### Motivation for this change

Build prometheus C library which I need for something that depends on it.

###### Things done

It's simple derivation except that it requires a one character patch. 

It's an addition, not a change, so it does not break anything. 

- Built on platform(s)
   - [ ] NixOS
   - [X] macOS
   - [X] other Linux distributions

- [X] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [X] Ensured that relevant documentation is up to date
- [X] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

### CODE REVIEW

File: pkgs/development/libraries/libprom/default.nix [hunk 1]
```diff
@@ -0,0 +1,39 @@
...
+    cd prom
+    export TEST=0
+  '';
+
+  meta = {
+    homepage = "https://github.com/digitalocean/prometheus-client-c/";
+    description = "A Prometheus Client in C";
+    platforms = stdenv.lib.platforms.unix;
+    license = stdenv.lib.licenses.asl20;
+    maintainers = [ maintainers.cfsmp3 ]
```
Comment (cole-h) [inline pkgs/development/libraries/libprom/default.nix:36]:
```suggestion
    maintainers = [ maintainers.cfsmp3 ];
```

File: pkgs/development/libraries/libprom/default.nix [hunk 2]
```diff
@@ -0,0 +1,39 @@
+{ gccStdenv
+, fetchFromGitHub
+, cmake
+}:
+let
+  stdenv = gccStdenv;
```
Comment (cole-h) [inline pkgs/development/libraries/libprom/default.nix:6]: Why is this necessary? I tested with the default `stdenv`, and it built fine.

File: pkgs/development/libraries/libprom/default.nix [hunk 3]
```diff
@@ -0,0 +1,39 @@
...
+in
+stdenv.mkDerivation rec {
+  pname = "libprom";
+  version = "0.1.1";
+
+  src = fetchFromGitHub {
+    rev = "v${version}";
+    owner = "digitalocean";
+    repo = "prometheus-client-c";
+    sha256 = "0g69s24xwrv5974acshrhnp6i8rpby8c6bhz15m3d8kpgjw3cm8f";
```
Comment (cole-h) [inline pkgs/development/libraries/libprom/default.nix:16]: Very minor nit (reordering to how most other packages organize this):
```suggestion
    owner = "digitalocean";
    repo = "prometheus-client-c";
    rev = "v${version}";
    sha256 = "0g69s24xwrv5974acshrhnp6i8rpby8c6bhz15m3d8kpgjw3cm8f";
```

File: pkgs/development/libraries/libprom/default.nix [hunk 4]
```diff
@@ -0,0 +1,39 @@
...
+  };
+
+  nativeBuildInputs = [ cmake ];
+  doCheck = false;
+
+  patches = [
+    ./prometheus-client-c-0.1.1.patch
```
Comment (cole-h) [inline pkgs/development/libraries/libprom/default.nix:23]: Could you please:

1) Add a comment why this patch is necessary (although I can vaguely guess)

2) Rename it to be more descriptive (e.g. "./fix-install-dir" or something) -- `prometheus-client-c-0.1.1.patch` tells me nothing about its contents, and it's nice for people to be able to skim a derivation and get a rough handle on what it's doing.

File: pkgs/development/libraries/libprom/default.nix [hunk 5]
```diff
@@ -0,0 +1,39 @@
...
+    cd prom
+    export TEST=0
+  '';
+
+  meta = with stdenv.lib; {
+    homepage = "https://github.com/digitalocean/prometheus-client-c/";
+    description = "A Prometheus Client in C";
+    platforms = stdenv.lib.platforms.unix;
+    license = stdenv.lib.licenses.asl20;
+    maintainers = [ maintainers.cfsmp3 ];
```
Comment (cole-h) [inline pkgs/development/libraries/libprom/default.nix:36]: You already use `stdenv.lib.*` almost everywhere in here, so might as well drop it from up here and use it actually-everywhere:

```suggestion
  meta = {
    homepage = "https://github.com/digitalocean/prometheus-client-c/";
    description = "A Prometheus Client in C";
    platforms = stdenv.lib.platforms.unix;
    license = stdenv.lib.licenses.asl20;
    maintainers = [ stdenv.lib.maintainers.cfsmp3 ];
```

File: pkgs/development/libraries/libprom/default.nix [hunk 6]
```diff
@@ -0,0 +1,39 @@
...
+  nativeBuildInputs = [ cmake ];
+  doCheck = false;
+
+  patches = [
+    ./prometheus-client-c-0.1.1.patch
+  ];
+
+  preConfigure = ''
+    cd prom
+    export TEST=0
```
Comment (cole-h) [inline pkgs/development/libraries/libprom/default.nix:28]: I know it's totally not your fault, but I *REALLY* hate the necessity of this export... More a question for upstream, but why not default to 0 if it's unset?!

**Review by cole-h — CHANGES_REQUESTED** [review]
Some changes I'd like to see, and a question.

File: pkgs/development/libraries/libprom/default.nix (hunk 2 above)
Comment (cfsmp3) [reply pkgs/development/libraries/libprom/default.nix:6]: So it builds on Mac.

File: pkgs/development/libraries/libprom/default.nix (hunk 4 above)
Comment (cfsmp3) [reply pkgs/development/libraries/libprom/default.nix:23]: Done.

File: pkgs/development/libraries/libprom/default.nix (hunk 5 above)
Comment (cfsmp3) [reply pkgs/development/libraries/libprom/default.nix:36]: Done.

File: pkgs/development/libraries/libprom/default.nix (hunk 6 above)
Comment (cfsmp3) [reply pkgs/development/libraries/libprom/default.nix:30]: Agreed to the hate. I'll probably send them a PR if/when they accept the one I sent already so we can get rid of that patch which is even worse than the export.

File: pkgs/development/libraries/libprom/default.nix (hunk 2 above)
Comment (cole-h) [reply pkgs/development/libraries/libprom/default.nix:6]: Understood. Maybe add a comment for that as well, then, please :)

File: pkgs/development/libraries/libprom/default.nix [hunk 7]
```diff
@@ -0,0 +1,41 @@
...
+    sha256 = "0g69s24xwrv5974acshrhnp6i8rpby8c6bhz15m3d8kpgjw3cm8f";
+  };
+
+  nativeBuildInputs = [ cmake ];
+  doCheck = false;
+
+  patches = [
+     # Required so CMAKE_INSTALL_PREFIX is honored, otherwise it
+     # installs headers in /usr/include (absolute)
+    ./fix-install-dir.patch
```
Comment (jtojnar) [inline pkgs/development/libraries/libprom/default.nix:25]: Is there an upstream issue or pull request? Perhaps suggest [`GNUInstallDirs`](https://cmake.org/cmake/help/v3.0/module/GNUInstallDirs.html) module.

File: pkgs/development/libraries/libprom/fix-install-dir.patch [hunk 8]
```diff
@@ -0,0 +1,10 @@
+diff -Naur prometheus-client-c-0.1.1.old/prom/CMakeLists.txt prometheus-client-c-0.1.1.new/prom/CMakeLists.txt
+--- prometheus-client-c-0.1.1.old/prom/CMakeLists.txt	2020-06-18 01:03:10.133059027 +0000
++++ prometheus-client-c-0.1.1.new/prom/CMakeLists.txt	2020-06-18 01:03:43.961228664 +0000
+@@ -137,4 +137,4 @@
+ 
+ include(CPack)
+ install(TARGETS prom ARCHIVE)
+-install(DIRECTORY include/ DESTINATION /usr/include/)
+\ No newline at end of file
++install(DIRECTORY include/ DESTINATION usr/include)
```
Comment (jtojnar) [inline pkgs/development/libraries/libprom/fix-install-dir.patch:10]:
```suggestion
+install(DIRECTORY include/ DESTINATION include)
```

or better with `GNUInstallDirs`

```suggestion
+install(DIRECTORY include/ DESTINATION ${CMAKE_INSTALL_FULL_INCLUDEDIR})
```

File: pkgs/development/libraries/libprom/default.nix (hunk 7 above)
Comment (cfsmp3) [reply pkgs/development/libraries/libprom/default.nix:25]: @jtojnar https://github.com/digitalocean/prometheus-client-c/pull/20

File: pkgs/development/libraries/libprom/fix-install-dir.patch (hunk 8 above)
Comment (cfsmp3) [reply pkgs/development/libraries/libprom/fix-install-dir.patch]: Done (the one without /usr ). I like the GNUInstallDirs option though, I'll update my PR to digitalocean.

Comment (jtojnar) [reply pkgs/development/libraries/libprom/fix-install-dir.patch]: I think it would be best to update the PR and then fetch the patch from there using `fetchpatch` (e.g. https://github.com/digitalocean/prometheus-client-c/commit/245c14aafde1dc015c6467f8145b5ffc4c3dbcf6.patch).

File: pkgs/development/libraries/libprom/default.nix [hunk 9]
```diff
@@ -0,0 +1,41 @@
+{ gccStdenv
+, fetchFromGitHub
+, cmake
+}:
+let
+  stdenv = gccStdenv; # So it builds on Mac
```
Comment (cole-h) [inline pkgs/development/libraries/libprom/default.nix:6]: Sorry, I meant a comment explaining *why* this fixes building on darwin -- e.g. what goes wrong without it.

On Linux, `stdenv` is `gccStdenv`. While it's possible, I don't see why Darwin wouldn't be the same.

Comment (cfsmp3) [reply pkgs/development/libraries/libprom/default.nix]: > On Linux, `stdenv` is `gccStdenv`. While it's possible, I don't see why Darwin wouldn't be the same.

Darwing defaults to clang which for some reason doesn't compile this.

Comment (purcell) [reply pkgs/development/libraries/libprom/default.nix]: Not sure either, but I was the one who wrote this part, having tested on MacOS and found that the default clang stdenv on Darwin did not work.

Comment (cole-h) [reply pkgs/development/libraries/libprom/default.nix]: Makes sense, thanks for the clarification (and for bearing with my nits).

**Review by cole-h — APPROVED** [review]
Diff LGTM, builds fine. Thanks for bearing with my nitpicking.

```
[2 built, 0.0 MiB DL]
https://github.com/NixOS/nixpkgs/pull/90706
1 package built:
libprom
```

**Comment by cfsmp3** [comment]
@jtojnar can you merge or approve that someone else does it? Thanks

**Comment by cole-h** [comment]
Before this gets merged, please squash your commits into a single `libprom: init at 0.1.1` commit.

File: pkgs/development/libraries/libprom/default.nix (hunk 7 above)
Comment (jtojnar) [reply pkgs/development/libraries/libprom/default.nix:25]: Also could you use `fetchpatch` with `https://github.com/digitalocean/prometheus-client-c/pull/20.patch`?

File: pkgs/development/libraries/libprom/default.nix (hunk 6 above)
Comment (jtojnar) [reply pkgs/development/libraries/libprom/default.nix:30]: Could you add a comment why this is needed? Looking at https://github.com/digitalocean/prometheus-client-c/blob/0c15e7e45ad0c3726593591fdd7d8f2fde845fe3/prom/CMakeLists.txt#L124-L126, I would expect this enables the tests but then you are not running them.

Comment (cole-h) [reply pkgs/development/libraries/libprom/default.nix:30]: If you don't have `TEST` set at all, building fails (checked locally when I made this comment).

**Comment by cfsmp3** [comment]
@cole-h Done.

File: pkgs/development/libraries/libprom/default.nix (hunk 6 above)
Comment (jtojnar) [reply pkgs/development/libraries/libprom/default.nix:30]: Hmm, perhaps `if ($ENV{TEST})` should be `if ("$ENV{TEST}")`.

**Comment by purcell** [comment]
Disk appears full on the machine running checks, hence the failure.

**Comment by cole-h** [comment]
Yep, we just dropped 3 old machines from ofborg that weren't cleaned up properly. Sorry about that.

@ofborg eval

**Comment by cfsmp3** [comment]
Thanks @cole-h , anything else needed from me?

**Comment by cole-h** [comment]
I think you missed https://github.com/NixOS/nixpkgs/pull/90706#discussion_r443717026. Aside from that, I don't see anything else wrong.

File: pkgs/development/libraries/libprom/default.nix (hunk 7 above)
Comment (cfsmp3) [reply pkgs/development/libraries/libprom/default.nix:25]: @jtojnar @cole-h That patch does not apply correctly to 0.1.1 I'm afraid.
```
applying patch /nix/store/karlwcmlzva7f9h9y90nqrjah07jg6yl-0c15e7e45ad0c3726593591fdd7d8f2fde845fe3.patch
patching file prom/CMakeLists.txt
Hunk #1 FAILED at 136.
1 out of 1 hunk FAILED -- saving rejects to file prom/CMakeLists.txt.rej
```
Since that patch upstream has been merged already maybe we can just merge this as is and completely remove patching as soon as they release a new version?

File: pkgs/development/libraries/libprom/fix-install-dir.patch (hunk 8 above)
Comment (cfsmp3) [reply pkgs/development/libraries/libprom/fix-install-dir.patch]: @jtojnar (sorry for the repeat message - just adding to the right conversation). That patch doesn't apply to 0.1.1.

File: pkgs/development/libraries/libprom/default.nix (hunk 7 above)
Comment (jtojnar) [reply pkgs/development/libraries/libprom/default.nix:25]: Could you also `fetchpatch` `https://github.com/digitalocean/prometheus-client-c/commit/5fcedeb506b7d47dd7bab35797f2c3f23db6fe10.patch`? Hopefully, that should make the patch apply and also resolve the `TEST` issue.

Comment (cfsmp3) [reply pkgs/development/libraries/libprom/default.nix]: Done.

File: pkgs/development/libraries/libprom/default.nix (hunk 6 above)
Comment (cfsmp3) [reply pkgs/development/libraries/libprom/default.nix]: This was fixed already upstream... I don't love this fetchpatch thing but well, operating under the assumption they'll release a new version soon and we'll be able to get rid of those patches.

**Comment by jtojnar** [comment]
Looks good. Thanks.
